### ISSUE TITLE
client stream write doesn't fail until 5 seconds elapsed after server finished

### ISSUE THREAD

**Issue by kuntryn** [grpc/grpc#33382]
### What version of gRPC and what language are you using?
grpc version: v1.55.0
language: c++

### What operating system (Linux, Windows,...) and version?
Linux

### What did you do?
For client streaming RPC, after server finished(closed)  the stream,  client's stream writing operation doesn't report  error (by returing false) immediately,  only after 5 seconds elapsed, the write operation  return false.

proto:
```c++
syntax = "proto3";

package com.test;

service FakeService{
    rpc clientStream(stream Request) returns (Response);
}

message Request{
    int32 i = 1;
}

message Response{
    string msg = 1;
}
``` 

server:
```c++
#include "test/proto/test.grpc.pb.h"
#include <grpcpp/grpcpp.h>

class FakeService : public com::test::FakeService::Service {
};

int main() {
  grpc::ServerBuilder builder;
  builder.AddListeningPort("127.0.0.1:50051",
                           grpc::InsecureServerCredentials());
  FakeService fake_service;
  builder.RegisterService(&fake_service);
  auto server = builder.BuildAndStart();

  server->Wait();
  return 0;
}
``` 

client
```c++
#include "test/proto/test.grpc.pb.h"
#include <grpcpp/grpcpp.h>

int main() {
  auto channel = grpc::CreateChannel("127.0.0.1:50051",
                                     grpc::InsecureChannelCredentials());
  auto stub = com::test::FakeService::NewStub(channel);

  grpc::ClientContext client_ctx;
  com::test::Response response;
  auto writer = stub->clientStream(&client_ctx, &response);

  for (int i = 0; true ; i++) {
    com::test::Request request;
    request.set_i(i);
    bool ok = writer->Write(request);
    if (!ok) {
      std::cout << i << "?" << std::endl;
      break;
    }
  }

  auto status = writer->Finish();
  if (!status.ok()) {
    std::cout << status.error_code() << " X " << status.error_message()
              << std::endl;
    return -1;
  }
  std::cout << "response msg: " << response.msg() << std::endl;

  return 0;
}
``` 
The client exits after 5 seconds.  The 5 seems related to this paramter from back_poller.
https://github.com/grpc/grpc/blob/a95e8c4ff46218ad7258dd2b22169c06c535050c/src/core/ext/filters/client_channel/backup_poller.cc#L41

### What did you expect to see?
Since the server's rpc method is unimplemented, the client's write operation should fail immediately

**Comment by yashykt** [comment]
The recommendation over here is to use the callback API. 

The issue that you are seeing is exactly alluded to the backup poller which will go away with the new event engine code. If you use the async callback API instead of the sync API you will just get that by default.

**Comment by yashykt** [comment]
This shouldn't be an issue normally with the sync API though. I suspect that the reason that we are not seeing polling over here is because there are no registered methods on the service, but let me verify that.

**Comment by yashykt** [comment]
So, I finally got the time to reproduce this on my machine. Turns out that what's happening is that since the client does not have an explicit read operation started by the application, the sync API does not have a thread that is actively polling the fd. This means that even though the server has responded with the error, the client does not actually poll the fd to get that read event until something like the backup poller comes in, which is why you are seeing the delay.

As I mentioned earlier, the sync API is known to have such issues, and hence our recommendations are switching to the callback API which doesn't have such pitfalls.
